File: source/bindings/java/com/microsoft/cognitiveservices/speech/dialog/SpeechBotConnector.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech.dialog;

import java.io.Closeable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import com.microsoft.cognitiveservices.speech.audio.AudioConfig;
import com.microsoft.cognitiveservices.speech.util.EventHandlerImpl;
import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.KeywordRecognitionModel;
import com.microsoft.cognitiveservices.speech.SessionEventArgs;
import com.microsoft.cognitiveservices.speech.SpeechConfig;
import com.microsoft.cognitiveservices.speech.SpeechRecognitionEventArgs;
import com.microsoft.cognitiveservices.speech.SpeechRecognitionCanceledEventArgs;
import com.microsoft.cognitiveservices.speech.translation.TranslationSynthesisEventArgs;

/**
 * Class that defines a SpeechBotConnector.
 */
public class SpeechBotConnector implements Closeable {

    /*! \cond PROTECTED */
    static Class<?> speechBotConnector = null;
    private static ExecutorService executorService;

    // load the native library.
    static {
        // trigger loading of native library
        try {
            Class.forName(SpeechConfig.class.getName());
        }
        catch (ClassNotFoundException ex) {
            throw new IllegalStateException(ex);
        }
        speechBotConnector = SpeechBotConnector.class;
        executorService = Executors.newCachedThreadPool();
    }

    /*! \endcond */

    /**
     * Builds a SpeechBotConnector with audio from default microphone input
     * @param config Bot connector configuration.
     */
    public SpeechBotConnector(BotConnectorConfig config) {
        this(config, AudioConfig.fromDefaultMicrophoneInput());
    }

    /**
     * Builds a SpeechBotConnector
     * @param config Bot connector configuration.
     * @param audioConfig An optional audio input configuration associated with the recognizer
     */
    public SpeechBotConnector(BotConnectorConfig config, AudioConfig audioConfig) {
        Contracts.throwIfNull(config, "config");
        if (audioConfig == null) {
            this.speechBotConnectorImpl = com.microsoft.cognitiveservices.speech.internal.SpeechBotConnector.FromConfig(config.getBotConfigImpl());
        } else {
            this.speechBotConnectorImpl = com.microsoft.cognitiveservices.speech.internal.SpeechBotConnector.FromConfig(config.getBotConfigImpl(), audioConfig.getConfigImpl());
        }
        initialize();
    }

    /**
     * Connects with the service.
     * @return A task representing the asynchronous operation that connects to the service.
     */
    public Future<Void> connectAsync() {
        return executorService.submit(new java.util.concurrent.Callable<Void>() {
            public Void call() {
                speechBotConnectorImpl.ConnectAsync().Get();
                return null;
            }
        });
    }

    /**
     * Disconnects from the service.
     * @return A task representing the asynchronous operation that disconnects from the service.
     */
    public Future<Void> disconnectAsync() {
        return executorService.submit(new java.util.concurrent.Callable<Void>() {
            public Void call() {
                speechBotConnectorImpl.DisconnectAsync().Get();
                return null;
            }
        });
    }

    /**
     * Sends an activity to the backing bot.
     * @param activity Activity to be sent.
     * @return A task representing the asynchronous operation that sends an activity to the bot.
     */
    public Future<String> sendActivityAsync(final BotConnectorActivity activity) {
        Contracts.throwIfNull(activity, "activity");
        return executorService.submit(new java.util.concurrent.Callable<String>() {
            public String call() {
                return speechBotConnectorImpl.SendActivityAsync(activity.getImpl()).Get();
            }
        });
    }

    /**
     * Starts a listening session that will terminate after the first utterance.
     * @return A task representing the asynchronous operation that starts a one shot listening session.
     */
    public Future<Void> listenOnceAsync() {
        return executorService.submit(new java.util.concurrent.Callable<Void>() {
            public Void call() {
                speechBotConnectorImpl.ListenOnceAsync().Get();
                return null;
            }
        });
    }

    /**
     * Initiates keyword recognition.
     * @param model The keyword recognition model that specifies the keyword to be recognized.
     * @return A task representing the asynchronous operation that starts keyword recognition.
     */
    public Future<Void> startKeywordRecognitionAsync(final KeywordRecognitionModel model) {
        return executorService.submit(new java.util.concurrent.Callable<Void>() {
            public Void call() {
                speechBotConnectorImpl.StartKeywordRecognitionAsync(model.getModelImpl()).Get();
                return null;
            }
        });
    }

    /**
     * Stop keyword recognition.
     * @return A task representing the asynchronous operation that stops keyword recognition.
     */
    public Future<Void> stopKeywordRecognitionAsync() {
        return executorService.submit(new java.util.concurrent.Callable<Void>() {
            public Void call() {
                speechBotConnectorImpl.StopKeywordRecognitionAsync().Get();
                return null;
            }
        });
    }

    /**
     * Defines event handler for the recognizing event.
     */
    public EventHandlerImpl<SpeechRecognitionEventArgs> recognizing = new EventHandlerImpl<SpeechRecognitionEventArgs>();

    /**
     * Defines event handler for the recognized event.
     */
    public EventHandlerImpl<SpeechRecognitionEventArgs> recognized = new EventHandlerImpl<SpeechRecognitionEventArgs>();

    /**
     * Defines event handler for the session start event.
     */
    public EventHandlerImpl<SessionEventArgs> sessionStarted = new EventHandlerImpl<SessionEventArgs>();

    /**
     * Defines event handler for the session stop event.
     */
    public EventHandlerImpl<SessionEventArgs> sessionStopped = new EventHandlerImpl<SessionEventArgs>();

    /**
     * Defines event handler for the canceled event.
     */
    public EventHandlerImpl<SpeechRecognitionCanceledEventArgs> canceled = new EventHandlerImpl<SpeechRecognitionCanceledEventArgs>();

    /**
     * Defines event handler for the activity received event.
     */
    public EventHandlerImpl<ActivityReceivedEventArgs> activityReceived = new EventHandlerImpl<ActivityReceivedEventArgs>();

    /*! \cond PROTECTED */
    private void initialize() {
        recognizingHandler = new RecoEventHandlerImpl(this, /* isRecognizedHandler */ false);
        speechBotConnectorImpl.getRecognizing().AddEventListener(recognizingHandler);
        recognizedHandler = new RecoEventHandlerImpl(this, /* isRecognizedHandler */ true);
        speechBotConnectorImpl.getRecognized().AddEventListener(recognizedHandler);
        sessionStartedHandler = new SessionEventHandlerImpl(this, /* isSessionStart */ true);
        speechBotConnectorImpl.getSessionStarted().AddEventListener(sessionStartedHandler);
        sessionStoppedHandler = new SessionEventHandlerImpl(this, /* isSessionStopped */ false);
        speechBotConnectorImpl.getSessionStopped().AddEventListener(sessionStoppedHandler);
        canceledHandler = new CanceledEventHandlerImpl(this);
        speechBotConnectorImpl.getCanceled().AddEventListener(canceledHandler);
        activityReceivedHandler = new ActivityReceivedEventHandlerImpl(this);
        speechBotConnectorImpl.getActivityReceived().AddEventListener(activityReceivedHandler);
    }

    private RecoEventHandlerImpl recognizingHandler;
    private RecoEventHandlerImpl recognizedHandler;
    private SessionEventHandlerImpl sessionStartedHandler;
    private SessionEventHandlerImpl sessionStoppedHandler;
    private CanceledEventHandlerImpl canceledHandler;
    private ActivityReceivedEventHandlerImpl activityReceivedHandler;

    private class RecoEventHandlerImpl extends com.microsoft.cognitiveservices.speech.internal.SpeechRecognitionEventListener {

        RecoEventHandlerImpl(SpeechBotConnector connector, boolean isRecognizedHandler) {
            Contracts.throwIfNull(connector, "connector");
            this.connector = connector;
            this.isRecognizedHandler = isRecognizedHandler;
        }

        @Override
        public void Execute(com.microsoft.cognitiveservices.speech.internal.SpeechRecognitionEventArgs eventArgs) {
            Contracts.throwIfNull(eventArgs, "eventArgs");
            if (connector.disposed) {
                return;
            }
            SpeechRecognitionEventArgs resultEventArg = new SpeechRecognitionEventArgs(eventArgs);
            EventHandlerImpl<SpeechRecognitionEventArgs> handler = this.isRecognizedHandler ? this.connector.recognized : this.connector.recognizing;
            if (handler != null) {
                handler.fireEvent(this.connector, resultEventArg);
            }
        }

        private SpeechBotConnector connector;
        private boolean isRecognizedHandler;
    }

    private class SessionEventHandlerImpl extends com.microsoft.cognitiveservices.speech.internal.SessionEventListener {

        SessionEventHandlerImpl(SpeechBotConnector connector, boolean isSessionStart) {
            Contracts.throwIfNull(connector, "connector");
            this.connector = connector;
            this.isSessionStart = isSessionStart;
        }

        @Override
        public void Execute(com.microsoft.cognitiveservices.speech.internal.SessionEventArgs eventArgs) {
            Contracts.throwIfNull(eventArgs, "eventArgs");
            if (connector.disposed) {
                return;
            }
            SessionEventArgs resultEventArg = new SessionEventArgs(eventArgs);
            EventHandlerImpl<SessionEventArgs> handler = this.isSessionStart ? this.connector.sessionStarted : this.connector.sessionStopped;
            if (handler != null) {
                handler.fireEvent(this.connector, resultEventArg);
            }
        }

        private SpeechBotConnector connector;
        private boolean isSessionStart;
    }

    private class CanceledEventHandlerImpl extends com.microsoft.cognitiveservices.speech.internal.SpeechRecognitionCanceledEventListener {

        CanceledEventHandlerImpl(SpeechBotConnector connector) {
            Contracts.throwIfNull(connector, "connector");
            this.connector = connector;
        }

        @Override
        public void Execute(com.microsoft.cognitiveservices.speech.internal.SpeechRecognitionCanceledEventArgs eventArgs) {
            Contracts.throwIfNull(eventArgs, "eventArgs");
            if (connector.disposed) {
                return;
            }
            SpeechRecognitionCanceledEventArgs resultEventArg = new SpeechRecognitionCanceledEventArgs(eventArgs);
            EventHandlerImpl<SpeechRecognitionCanceledEventArgs> handler = this.connector.canceled;
            if (handler != null) {
                handler.fireEvent(this.connector, resultEventArg);
            }
        }

        private SpeechBotConnector connector;
    }

    private class ActivityReceivedEventHandlerImpl extends com.microsoft.cognitiveservices.speech.internal.ActivityReceivedEventListener {

        ActivityReceivedEventHandlerImpl(SpeechBotConnector connector) {
            Contracts.throwIfNull(connector, "connector");
            this.connector = connector;
        }

        @Override
        public void Execute(com.microsoft.cognitiveservices.speech.internal.ActivityReceivedEventArgs eventArgs) {
            Contracts.throwIfNull(eventArgs, "eventArgs");
            if (connector.disposed) {
                return;
            }
            ActivityReceivedEventArgs resultEventArg = new ActivityReceivedEventArgs(eventArgs);
            EventHandlerImpl<ActivityReceivedEventArgs> handler = this.connector.activityReceived;
            if (handler != null) {
                handler.fireEvent(this.connector, resultEventArg);
            }
        }

        private SpeechBotConnector connector;
    }

    private com.microsoft.cognitiveservices.speech.internal.SpeechBotConnector speechBotConnectorImpl;
    /*! \endcond */

    /**
     * Dispose of associated resources.
     */
    @Override
    public void close() {
        dispose(true);
    }

    /*! \cond PROTECTED */
    private boolean disposed = false;

    protected void dispose(boolean disposing) {
        if (disposed) {
            return;
        }
        if (disposing) {
            speechBotConnectorImpl.getRecognizing().RemoveEventListener(recognizingHandler);
            speechBotConnectorImpl.getRecognized().RemoveEventListener(recognizedHandler);
            speechBotConnectorImpl.getSessionStarted().RemoveEventListener(sessionStartedHandler);
            speechBotConnectorImpl.getSessionStopped().RemoveEventListener(sessionStoppedHandler);
            speechBotConnectorImpl.getCanceled().RemoveEventListener(canceledHandler);
            speechBotConnectorImpl.getActivityReceived().RemoveEventListener(activityReceivedHandler);

            recognizingHandler.delete();
            recognizedHandler.delete();
            sessionStartedHandler.delete();
            sessionStoppedHandler.delete();
            canceledHandler.delete();
            activityReceivedHandler.delete();

            speechBotConnectorImpl.delete();
        }
        disposed = true;
    }
    /*! \endcond */
}

File: source/bindings/csharp/speech_bot_connector.cs
//
// Copyright (c) Microsoft. All rights reserved.
// See https://aka.ms/csspeech/license201809 for the full license information.
//

using System;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using Microsoft.CognitiveServices.Speech.Internal;
using Microsoft.CognitiveServices.Speech.Translation;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;

namespace Microsoft.CognitiveServices.Speech.Dialog
{
    /// <summary>
    /// Connects to a speech enabled bot.
    /// Added in version 1.5.0
    /// </summary>
    public sealed class SpeechBotConnector : IDisposable
    {
        /// <summary>
        /// Signal that indicates the start of a listening session.
        /// </summary>
        public event EventHandler<SessionEventArgs> SessionStarted;

        /// <summary>
        /// Signal that indicates the end of a listening session.
        /// </summary>
        public event EventHandler<SessionEventArgs> SessionStopped;

        /// <summary>
        /// Signal for events containing speech recognition results.
        /// </summary>
        public event EventHandler<SpeechRecognitionEventArgs> Recognized;

        /// <summary>
        /// Signal for events containing intermediate recognition results.
        /// </summary>
        public event EventHandler<SpeechRecognitionEventArgs> Recognizing;

        /// <summary>
        /// Signal for events relating to the cancellation of an interaction.
        /// </summary>
        public event EventHandler<SpeechRecognitionCanceledEventArgs> Canceled;

        /// <summary>
        /// Signal that an activity was received from the bot.
        /// </summary>
        public event EventHandler<ActivityReceivedEventArgs> ActivityReceived;

        /// <summary>
        /// Creates a speech bot connector using the default microphone input for a specified bot connector configuration.
        /// </summary>
        /// <param name="config">Bot connector config.</param>
        /// <returns>A speech bot connector instance.</returns>
        public SpeechBotConnector(BotConnectorConfig config)
            : this(FromConfig(SpxFactory.bot_connector_create_speech_bot_connector_from_config, config))
        {
        }

        /// <summary>
        /// Creates a speech bot connector using the specified bot connector and audio configuration.
        /// </summary>
        /// <param name="config">Bot connector config.</param>
        /// <param name="audioConfig">Audio config.</param>
        /// <returns>A speech bot connector instance.</returns>
        public SpeechBotConnector(BotConnectorConfig config, Audio.AudioConfig audioConfig)
            : this(FromConfig(SpxFactory.bot_connector_create_speech_bot_connector_from_config, config, audioConfig))
        {
        }

        internal SpeechBotConnector(InteropSafeHandle botPtr)
        {
            ThrowIfNull(botPtr);
            botConnectorHandle = botPtr;
            gch = GCHandle.Alloc(this, GCHandleType.Weak);

            sessionStartedCallbackDelegate = FireEvent_SessionStarted;
            sessionStoppedCallbackDelegate = FireEvent_SessionStopped;
            recognizedCallbackDelegate = FireEvent_Recognized;
            recognizingCallbackDelegate = FireEvent_Recognizing;
            canceledCallbackDelegate = FireEvent_Canceled;
            activityReceivedCallbackDelegate = FireEvent_ActivityReceived;

            ThrowIfFail(Internal.SpeechBotConnector.bot_connector_session_started_set_callback(botConnectorHandle, sessionStartedCallbackDelegate, GCHandle.ToIntPtr(gch)));
            ThrowIfFail(Internal.SpeechBotConnector.bot_connector_session_stopped_set_callback(botConnectorHandle, sessionStoppedCallbackDelegate, GCHandle.ToIntPtr(gch)));
            ThrowIfFail(Internal.SpeechBotConnector.bot_connector_recognized_set_callback(botConnectorHandle, recognizedCallbackDelegate, GCHandle.ToIntPtr(gch)));
            ThrowIfFail(Internal.SpeechBotConnector.bot_connector_recognizing_set_callback(botConnectorHandle, recognizingCallbackDelegate, GCHandle.ToIntPtr(gch)));
            ThrowIfFail(Internal.SpeechBotConnector.bot_connector_canceled_set_callback(botConnectorHandle, canceledCallbackDelegate, GCHandle.ToIntPtr(gch)));
            ThrowIfFail(Internal.SpeechBotConnector.bot_connector_activity_received_set_callback(botConnectorHandle, activityReceivedCallbackDelegate, GCHandle.ToIntPtr(gch)));
        }

        ~SpeechBotConnector()
        {
            isDisposing = true;
            Dispose(false);
        }

        /// <summary>
        /// Dispose of associated resources.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// This method performs cleanup of resources.
        /// The Boolean parameter <paramref name="disposing"/> indicates whether the method is called from <see cref="IDisposable.Dispose"/> (if <paramref name="disposing"/> is true) or from the finalizer (if <paramref name="disposing"/> is false).
        /// </summary>
        /// <param name="disposing">Flag to request disposal.</param>
        internal void Dispose(bool disposing)
        {
            if (disposed)
            {
                return;
            }

            if (botConnectorHandle != null)
            {
                LogErrorIfFail(Internal.SpeechBotConnector.bot_connector_session_started_set_callback(botConnectorHandle, null, IntPtr.Zero));
                LogErrorIfFail(Internal.SpeechBotConnector.bot_connector_session_stopped_set_callback(botConnectorHandle, null, IntPtr.Zero));
                LogErrorIfFail(Internal.SpeechBotConnector.bot_connector_recognized_set_callback(botConnectorHandle, null, IntPtr.Zero));
                LogErrorIfFail(Internal.SpeechBotConnector.bot_connector_recognizing_set_callback(botConnectorHandle, null, IntPtr.Zero));
                LogErrorIfFail(Internal.SpeechBotConnector.bot_connector_canceled_set_callback(botConnectorHandle, null, IntPtr.Zero));
                LogErrorIfFail(Internal.SpeechBotConnector.bot_connector_activity_received_set_callback(botConnectorHandle, null, IntPtr.Zero));
            }

            // Dispose of managed resources
            if (disposing)
            {
                botConnectorHandle.Dispose();
            }

            sessionStartedCallbackDelegate = null;
            sessionStoppedCallbackDelegate = null;
            recognizedCallbackDelegate = null;
            recognizingCallbackDelegate = null;
            canceledCallbackDelegate = null;
            activityReceivedCallbackDelegate = null;

            // Release any unmanaged resources not wrapped by safe handles
            if (gch.IsAllocated)
            {
                gch.Free();
            }
            disposed = true;
        }

        internal InteropSafeHandle botConnectorHandle;
        private IntPtr asyncStartContinuousHandle = IntPtr.Zero;
        private IntPtr asyncStopContinuousHandle = IntPtr.Zero;
        private IntPtr asyncStartKeywordHandle = IntPtr.Zero;
        private IntPtr asyncStopKeywordHandle = IntPtr.Zero;
        private CallbackFunctionDelegate sessionStartedCallbackDelegate;
        private CallbackFunctionDelegate sessionStoppedCallbackDelegate;
        private CallbackFunctionDelegate recognizedCallbackDelegate;
        private CallbackFunctionDelegate recognizingCallbackDelegate;
        private CallbackFunctionDelegate canceledCallbackDelegate;
        private CallbackFunctionDelegate activityReceivedCallbackDelegate;

        /// <summary>
        /// GC handle for callbacks for context.
        /// </summary>
        private GCHandle gch;

        /// <summary>
        /// disposed is a flag used to indicate if object is disposed.
        /// </summary>
        private volatile bool disposed = false;

        /// <summary>
        /// isDisposing is a flag used to indicate if object is being disposed.
        /// </summary>
        private volatile bool isDisposing = false;

        /// <summary>
        /// botLock is used to synchronize access to objects member variables from multiple threads
        /// </summary>
        private readonly object botLock = new object();

        static SpeechBotConnector GetConnectorFromContext(IntPtr context)
        {
            var connector = InteropSafeHandle.GetObjectFromWeakHandle<SpeechBotConnector>(context);
            if (connector == null || connector.isDisposing)
            {
                return null;
            }
            return connector;
        }

        static void FireEvent<EvtArgs>(EvtArgs eventArgs, SpeechBotConnector connector, EventHandler<EvtArgs> evt)
        {
            evt?.Invoke(connector, eventArgs);
        }

        /// <summary>
        /// Method to raise a SessionStarted C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        [MonoPInvokeCallback]
        private static void FireEvent_SessionStarted(IntPtr handle, IntPtr eventHandle, IntPtr context)
        {
            try
            {
                var connector = GetConnectorFromContext(context);
                FireEvent(new SessionEventArgs(eventHandle), connector, connector?.SessionStarted);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        /// <summary>
        /// Method to raise a SessionStopped C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        [MonoPInvokeCallback]
        private static void FireEvent_SessionStopped(IntPtr handle, IntPtr eventHandle, IntPtr context)
        {
            try
            {
                var connector = GetConnectorFromContext(context);
                FireEvent(new SessionEventArgs(eventHandle), connector, connector?.SessionStopped);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        /// <summary>
        /// Method to raise a Recognizing C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        [MonoPInvokeCallback]
        private static void FireEvent_Recognizing(IntPtr handle, IntPtr eventHandle, IntPtr context)
        {
            try
            {
                var connector = GetConnectorFromContext(context);
                FireEvent(new SpeechRecognitionEventArgs(eventHandle), connector, connector?.Recognizing);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        /// <summary>
        /// Method to raise a Recognized C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        [MonoPInvokeCallback]
        private static void FireEvent_Recognized(IntPtr handle, IntPtr eventHandle, IntPtr context)
        {
            try
            {
                var connector = GetConnectorFromContext(context);
                FireEvent(new SpeechRecognitionEventArgs(eventHandle), connector, connector?.Recognized);
            }
            catch (InvalidOperationException)
            {
                SpxExceptionThrower.LogError(SpxError.InvalidHandle);
            }
        }


        /// <summary>
        /// Method to raise a Canceled C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        [MonoPInvokeCallback]
        private static void FireEvent_Canceled(IntPtr handle, IntPtr eventHandle, IntPtr context)
        {
            try
            {
                var connector = GetConnectorFromContext(context);
                FireEvent(new SpeechRecognitionCanceledEventArgs(eventHandle), connector, connector?.Canceled);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }


        /// <summary>
        /// Method to raise a ActivityReceived C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        [MonoPInvokeCallback]
        private static void FireEvent_ActivityReceived(IntPtr handle, IntPtr eventHandle, IntPtr context)
        {
            try
            {
                var connector = GetConnectorFromContext(context);
                FireEvent(new ActivityReceivedEventArgs(eventHandle), connector, connector?.ActivityReceived);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        private void AssertNotDisposed()
        {
            if (disposed || isDisposing)
            {
                throw new ObjectDisposedException(this.GetType().Name);
            }
        }

        /// <summary>
        /// Connects with the back end.
        /// </summary>
        /// <returns>An asynchronous operation that starts the connection.</returns>
        public Task ConnectAsync()
        {
            AssertNotDisposed();

            return Task.Run(() =>
            {
                ThrowIfNull(botConnectorHandle, "Invalid connector handle");
                ThrowIfFail(Internal.SpeechBotConnector.bot_connector_connect(botConnectorHandle));
            });
        }

        /// <summary>
        /// Disconnects from the back end.
        /// </summary>
        /// <returns>An asynchronous operation that starts the disconnection.</returns>
        public Task DisconnectAsync()
        {
            AssertNotDisposed();

            return Task.Run(() =>
            {
                ThrowIfNull(botConnectorHandle, "Invalid connector handle");
                ThrowIfFail(Internal.SpeechBotConnector.bot_connector_disconnect(botConnectorHandle));
            });
        }

        /// <summary>
        /// Sends an activity to the backing bot.
        /// </summary>
        /// <param name="activityJSON">Activity to send as a serialized JSON</param>
        /// <returns>An asynchronous operation that starts the operation.</returns>
        public Task<string> SendActivityAsync(string activityJSON)
        {
            AssertNotDisposed();

            return Task.Run(() =>
            {
                IntPtr activityPtr = IntPtr.Zero;
                ThrowIfNull(botConnectorHandle, "Invalid connector handle");
                ThrowIfFail(Internal.BotConnectorActivity.bot_activity_from_string(activityJSON, out activityPtr));
                const int guidSize = 50;
                var buffer = new StringBuilder(guidSize);
                ThrowIfFail(Internal.SpeechBotConnector.bot_connector_send_activity(botConnectorHandle, activityPtr, buffer));
                ThrowIfFail(Internal.BotConnectorActivity.bot_activity_handle_release(activityPtr));
                return buffer.ToString();
            });
        }

        /// <summary>
        /// Initiates keyword recognition.
        /// </summary>
        /// <param name="model">Specifies the keyword model to be used.</param>
        /// <returns>An asynchronous operation that starts the operation.</returns>
        public Task StartKeywordRecognitionAsync(KeywordRecognitionModel model)
        {
            AssertNotDisposed();

            ThrowIfNull(botConnectorHandle, "Invalid connector handle");
            return Task.Run(() =>
            {
                ThrowIfNull(botConnectorHandle, "Invalid connector handle");
                ThrowIfFail(Internal.SpeechBotConnector.bot_connector_start_keyword_recognition(botConnectorHandle, model.keywordHandle));
            });
        }

        /// <summary>
        /// Stop keyword recognition.
        /// </summary>
        /// <returns>An asynchronous operation that starts the operation.</returns>
        public Task StopKeywordRecognitionAsync()
        {
            AssertNotDisposed();

            return Task.Run(() =>
            {
                ThrowIfNull(botConnectorHandle, "Invalid connector handle");
                ThrowIfFail(Internal.SpeechBotConnector.bot_connector_stop_keyword_recognition(botConnectorHandle));
            });
        }

        /// <summary>
        /// Starts a listening session that will terminate after the first utterance.
        /// </summary>
        /// <returns>An asynchronous operation that starts the operation.</returns>
        public Task ListenOnceAsync()
        {
            AssertNotDisposed();

            return Task.Run(() =>
            {
                ThrowIfNull(botConnectorHandle, "Invalid connector handle");
                ThrowIfFail(Internal.SpeechBotConnector.bot_connector_listen_once(botConnectorHandle));
            });
        }

        internal delegate IntPtr GetRecognizerFromConfigDelegate(out IntPtr phreco, InteropSafeHandle speechconfig, InteropSafeHandle audioInput);

        internal static InteropSafeHandle FromConfig(GetRecognizerFromConfigDelegate fromConfig, SpeechConfig speechConfig, Audio.AudioConfig audioConfig)
        {
            if (speechConfig == null) throw new ArgumentNullException(nameof(speechConfig));
            if (audioConfig == null) throw new ArgumentNullException(nameof(audioConfig));

            IntPtr recoHandlePtr = IntPtr.Zero;
            ThrowIfFail(fromConfig(out recoHandlePtr, speechConfig.configHandle, audioConfig.configHandle));
            InteropSafeHandle recoHandle = new InteropSafeHandle(recoHandlePtr, Internal.SpeechBotConnector.bot_connector_handle_release);
            GC.KeepAlive(speechConfig);
            GC.KeepAlive(audioConfig);
            return recoHandle;
        }

        internal static InteropSafeHandle FromConfig(GetRecognizerFromConfigDelegate fromConfig, SpeechConfig speechConfig)
        {
            if (speechConfig == null) throw new ArgumentNullException(nameof(speechConfig));

            IntPtr recoHandlePtr = IntPtr.Zero;
            IntPtr audioConfigPtr = IntPtr.Zero;
            InteropSafeHandle audioConfigHandle = new InteropSafeHandle(audioConfigPtr, null);
            ThrowIfFail(fromConfig(out recoHandlePtr, speechConfig.configHandle, audioConfigHandle));
            InteropSafeHandle recoHandle = new InteropSafeHandle(recoHandlePtr, Internal.SpeechBotConnector.bot_connector_handle_release);
            GC.KeepAlive(speechConfig);
            return recoHandle;
        }


    }
}
